docs/Security_Enhancement_For_AI_Models.md

```diff
@@ -0,0 +1,246 @@
+
+- Deploy CCNP as a [DaemonSet](https://intel.github.io/confidential-cloud-native-primitives/_rst/service.quote.html#deploy-as-daemonset-in-kubernetes), it will detect the TEE environment.
+- Get the quote and request the key from KBS.
+- Get the encrypted model from model server.
+- Decrypt the model and load the model do inference.
+
+## 2. Implementation
+
+Current Cloud Native AI Pipeline (CNAP) already implemented the new design to protect the models, this section will describe the details of each components and introduce how to enable a new model server or KBS interface.
+
+To protect the model, the first thing is generating a key then encypting the model and register the key to KBS, KBS could submit request to store the key to KMS.
+To get a model and decrypt the model, the runtime environment should be verified by attestation. A key broker client should get a quote and send the quote to KBS to attest the environment is TEE then response a wrapped key for decryption.
+
+### 2.1 Mode Encryption
+
+AES GCM is a widely used encryption algorithm and mode of operation for securing data, particularly in network communications and data storage. It combines two important cryptographic techniques: AES and GCM. AES is a symmetric encryption algorithm that uses a shared secret key for both encryption and decryption. GCM is based on a combination of CTR and GMAC, this combination ensures that not only is the data encrypted, but it also includes an authentication tag that allows the recipient to verify the integrity of the data.
+
+The key features and benefits of AES GCM include confidentiality, data integrity and authentication. The Initialization Vector (IV) is a random value used as an input alongside the encryption key to initialize the encryption process, and the authentication tag is a small piece of data generated during encryption and appended to the ciphertext.
+
+Typically, the AES 256 GCM encrypted data format should be:
+
+12 bytes header | [12] bytes IV | encrypted data | [16] bytes tag
+---|---|---|---
+
+and the 12 bytes header:
+
+uint32 IV length | uint32 tag length | uint32 data length
+---|---|---
+
+```
+
+Some sample codes for encyrption:
+
+```Python
+import secrets
+import struct
+from cryptography.hazmat.primitives.ciphers.aead import AESGCM
+
+def encrypt_file(input, output, key):
+    with open(args.input, 'rb') as infile:
+        data = infile.read()
+        aesgcm = AESGCM(key)
+        IV = secrets.token_bytes(12)
+        data_enc = aesgcm.encrypt(IV, data, None)
+        with open(output, 'wb') as outfile:
+            header = struct.pack('<3I', len(IV), 16, len(data_enc))
+            outfile.write(header)
+            outfile.write(IV)
+            outfile.write(data_enc)
+```
+
+### 2.2 Key Registration
+
+After encrypting the model, the key should be KBS, a KBS shuold handle the attestation for a client to get the key. A typical flow of key registration:
```

Comment:
After encrypting the model, the key should be registered with KBS. The KBS handles the attestation process for a client to retrieve the key.

Reply:
Done.